File: 이재진/피로도.java
class Solution {
    public boolean[] visited;
    public int maxCount;
    
    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        maxCount = 0;
        dfs(k,dungeons,0);
        return maxCount;
    }
    
    public void dfs(int k, int[][] dungeons, int count) {
        maxCount = Math.max(maxCount, count);
        
        for (int i=0; i<dungeons.length; i++) {
            int req = dungeons[i][0];
            int use = dungeons[i][1];
            
            if (!visited[i] && k >= req) {
                visited[i] = true;
                dfs(k-use, dungeons, count+1);
                visited[i] = false;
            }
            
        }
    }
}

File: 정종길/프로그래머스/기타/피로도.py
answer = -1


def solution(k, dungeons):
    def dfs(k, now, check, cnt):
        global answer

        if k >= dungeons[now][0]:
            k -= dungeons[now][1]
            cnt += 1

        if check == len(dungeons) - 1:
            answer = max(answer, cnt)

        for i in range(len(dungeons)):
            if not visited[i]:
                visited[i] = 1
                dfs(k, i, check + 1, cnt)
                visited[i] = 0

    visited = [0] * len(dungeons)

    for i in range(len(dungeons)):
        visited[i] = 1
        dfs(k, i, 0, 0)
        visited[i] = 0

    return answer
